### ISSUE TITLE
AC charging appears as "Unknown"

### ISSUE THREAD

**Issue by transistor1** [Field-Effect-LLC/RDP-BattMon#17]
When the computer is plugged in, it has a BatteryStatus of 2, and thus appears as "Unknown" (at least on my machine).

However, according to:
https://msdn.microsoft.com/en-us/library/aa394074%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396

> **Unknown (2)**
> 
> The system has access to AC so no battery is being discharged. However, the battery is not necessarily charging.

**Comment by transistor1** [comment]
This isn't really a bug, because the battery is reporting a (2), and Microsoft says a 2 is Unknown, so I'm marking it as an Enhancement.

**Comment by peterberbec** [comment]
Very odd that there isn't more definition. But a simple "unknown = AC charge" won't work. See:

![image](https://user-images.githubusercontent.com/17226353/37249039-58a0cb84-24ad-11e8-8f0f-d75533203190.png)

I'm not on AC, so one battery is discharging and the other is unknown. You'd need to test every battery being unknown to be sure it is charging.

`bool isCharging()
{
	for(int i=0; i<NUM_BATTERIES; i++)
	{
		if(battery[i].getStatus() != UNKNOWN)
		{
			return false;
		}
	}
	return true;
}
`

**Comment by transistor1** [comment]
It's weird that you're not on AC, yet it still says "Unknown".

There's probably some magical incantation to determine whether it's on AC, or to figure out exactly what Unknown means.  It's probably some Microsofty combination of fields.

This is my first foray into the world of battery data, so I'm coming from a very naive place.  If you look at the complete description for "BatteryStatus" in the page I had linked to,  I essentially took the BatteryStatus enumeration at face value:

> BatteryStatus
> Data type: uint16
> Access type: Read-only
> Qualifiers: MappingStrings ("MIF.DMTF|Portable Battery|002.14")
> Status of the battery. The value 10 (Undefined) is not valid in the CIM schema because in DMI it represents that no battery is installed. In this case, the object should not be instantiated.
> 
> This property is inherited from CIM_Battery.
> 
> Other (1)
> 
> The battery is discharging.
> 
> Unknown (2)
> 
> The system has access to AC so no battery is being discharged. However, the battery is not necessarily charging.
> 
> Fully Charged (3)
> 
> Low (4)
> 
> Critical (5)
> 
> Charging (6)
> 
> Charging and High (7)
> 
> Charging and Low (8)
> 
> Charging and Critical (9)
> 
> Undefined (10)
> 
> Partially Charged (11)
>

**Comment by peterberbec** [comment]
Yeah they're obviously lying. We can see that (2) actually means not discharging.

If ALL batteries are NOT discharging, the computer IS charging.

**Comment by BobVul** [comment]
That's not a given; some laptops (e.g. Dell) will refuse to charge if they can't detect the AC adapter type. It's a fairly common problem both with generic chargers and when the sense line (tiny pin in the middle of the plug) breaks on an OEM adapter. 

In those cases, the computer will run off mains but not charge.

**Comment by peterberbec** [comment]
Interesting. So how would one determine the difference between:

A. All batteries not discharging, running off mains, charging 

and 

B. All batteries not discharging, running off mains, not charging?

**Comment by transistor1** [comment]
@BobVul that's a good point.  I had a 2006 Apple MacBook Pro that relied on a sensing pin & wouldn't charge when that was broken.

I was about to write that I'd implement your suggestion, @peterberbec , but we don't necessarily know the difference between "charging is broken Unknown" and "I'm really charging Unknown"

**Comment by transistor1** [comment]
[Possible solution](https://blogs.technet.microsoft.com/heyscriptingguy/2007/04/09/how-can-i-tell-whether-a-laptop-computer-is-running-off-batteries/) using WMI query:

VBScript:
```
strComputer = "."
Set objWMIService = GetObject("winmgmts:\\" & strComputer & "\root\wmi")
Set colItems = objWMIService.ExecQuery("Select * From BatteryStatus Where Voltage > 0")

For Each objItem in colItems
    Wscript.Echo "Battery: " & objItem.InstanceName
    Wscript.Echo "On AC Power: " & objItem.PowerOnline
    Wscript.Echo "Battery is Discharging: " & objItem.Discharging
    Wscript.Echo "Battery is Charging: " & objItem.Charging
    Wscript.Echo "Remaining capacity: " & objItem.RemainingCapacity
Next
```

**Comment by transistor1** [comment]
[More useful info on WMI classes](https://superuser.com/a/995048/94697)
